DuckDuckGo/JSAlertController.swift

```diff
@@ -0,0 +1,208 @@
+                self.present(alert)
+            }
+        }
+    }
+
+    private func reloadData() {
+        guard isViewLoaded,
+            let alert = alert
+        else { return }
+
+        okButton.setTitle(UserText.webJSAlertOKButton, for: .normal)
+        okButton.setBackgroundImage(.highlightedAlertButtonTint, for: .highlighted)
+        cancelButton.setTitle(UserText.webJSAlertCancelButton, for: .normal)
+        cancelButton.setBackgroundImage(.highlightedAlertButtonTint, for: .highlighted)
+        closeTabButton.setTitle(UserText.webJSAlertCloseTabButton, for: .normal)
+        closeTabButton.setBackgroundImage(.highlightedAlertButtonTint, for: .highlighted)
+
+        titleLabel.text = String(format: UserText.webJSAlertWebsiteMessageFormat, alert.domain)
+        messageLabel.text = alert.message
+
+        if let text = alert.text {
+            textField.placeholder = text
+            textField.text = text
+            textFieldBox.isHidden = false
+        } else {
+            textFieldBox.isHidden = true
+        }
+
+        cancelButton.isHidden = !alert.isConfirm
+        cancelButtonSeparator.isHidden = !alert.isConfirm
+    }
+
+    @IBAction func okAction(_ sender: UIButton) {
+        dismiss(animated: true) { [alert=self.alert, text=self.textField.text] in
+            alert?.complete(with: true, text: text)
+        }
+    }
+
+    @IBAction func cancelAction(_ sender: Any) {
+        dismiss(animated: true) { [alert=self.alert] in
+            alert?.complete(with: false, text: nil)
+        }
+    }
+
+    @IBAction func closeTabAction(_ sender: Any) {
+        dismiss(animated: true) { [alert=self.alert] in
+            alert?.closeTab()
+        }
+    }
+
+    private func registerForKeyboardNotifications() {
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(keyboardDidShow),
+                                               name: UIResponder.keyboardDidShowNotification,
+                                               object: nil)
+
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(keyboardWillHide),
+                                               name: UIResponder.keyboardWillHideNotification,
+                                               object: nil)
+    }
+
+    @objc private func keyboardDidShow(notification: NSNotification) {
+        guard let isLocalUserInfoKey = notification.userInfo?[UIResponder.keyboardIsLocalUserInfoKey] as? NSNumber,
+              isLocalUserInfoKey == true,
+              let intersection = (notification.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect)?
+                  .intersection(self.view.convert(view.bounds, to: view.window)),
+              self.keyboardConstraint.constant != intersection.height
+        else {
+            return
+        }
+
+        UIView.animate(withDuration: 0.3) {
```

Comment:
Maybe it would be good to create a private enum with all these animation durations?